drivers/usb/udc/udc_renesas_ra.c

```diff
@@ -0,0 +1,756 @@
+	UDC_RENESAS_RA_EVT_STATUS,
+};
+
+struct udc_renesas_ra_evt {
+	enum udc_renesas_ra_event_type type;
+	usbd_event_t hal_evt;
+	uint8_t ep;
+};
+
+K_MSGQ_DEFINE(drv_msgq, sizeof(struct udc_renesas_ra_evt), CONFIG_UDC_RENESAS_RA_MAX_QMESSAGES,
+	      sizeof(uint32_t));
+
+extern void usb_device_isr(void);
+
+static void udc_event_handler(usbd_callback_arg_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+	struct udc_renesas_ra_evt evt;
+
+	switch (p_args->event.event_id) {
+	case USBD_EVENT_BUS_RESET:
+		udc_submit_event(dev, UDC_EVT_RESET, 0);
+		break;
+
+	case USBD_EVENT_VBUS_RDY:
+		udc_submit_event(dev, UDC_EVT_VBUS_READY, 0);
+		break;
+
+	case USBD_EVENT_VBUS_REMOVED:
+		udc_submit_event(dev, UDC_EVT_VBUS_REMOVED, 0);
+		break;
+
+	case USBD_EVENT_SUSPEND:
+		udc_submit_event(dev, UDC_EVT_SUSPEND, 0);
+		break;
+
+	case USBD_EVENT_RESUME:
+		udc_submit_event(dev, UDC_EVT_RESUME, 0);
+		break;
+
+	case USBD_EVENT_SOF:
+		udc_submit_event(dev, UDC_EVT_SOF, 0);
+		break;
+
+	default:
+		evt.type = UDC_RENESAS_RA_EVT_HAL;
+		evt.hal_evt = p_args->event;
+		k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
+		break;
+	}
+}
+
+static void usb_interrupt_handler(void *arg)
+{
+	ARG_UNUSED(arg);
+	usb_device_isr();
+}
+
+static void udc_event_xfer_next(const struct device *dev, const uint8_t ep)
+{
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+	struct net_buf *buf;
+
+	if (udc_ep_is_busy(dev, ep)) {
+		return;
+	}
+
+	buf = udc_buf_peek(dev, ep);
+	if (buf != NULL) {
+		int err;
+
+		if (USB_EP_DIR_IS_IN(ep)) {
+			err = R_USBD_XferStart(&data->udc, ep, buf->data, buf->len);
+		} else {
+			err = R_USBD_XferStart(&data->udc, ep, buf->data, buf->size);
+		}
+
+		if (err != FSP_SUCCESS) {
+			LOG_ERR("ep 0x%02x error", ep);
+			/* REVISE: remove from endpoint queue? ASSERT? */
+			udc_submit_ep_event(dev, buf, -ECONNREFUSED);
+		} else {
+			udc_ep_set_busy(dev, ep, true);
+		}
+	}
+}
+
+static int usbd_ctrl_feed_dout(const struct device *dev, const size_t length)
+{
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	struct net_buf *buf;
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	k_fifo_put(&cfg->fifo, buf);
+
+	return FSP_SUCCESS == R_USBD_XferStart(&data->udc, cfg->addr, buf->data, buf->size) ? 0
+											    : -EIO;
+}
+
+static int udc_event_xfer_setup(const struct device *dev, struct udc_renesas_ra_evt *evt)
+{
+	struct net_buf *buf;
+	int err;
+
+	struct usb_setup_packet *setup_packet =
+		(struct usb_setup_packet *)&evt->hal_evt.setup_received;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, sizeof(struct usb_setup_packet));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate for setup");
+		return -ENOMEM;
+	}
+
+	udc_ep_buf_set_setup(buf);
+	memcpy(buf->data, setup_packet, sizeof(struct usb_setup_packet));
+	net_buf_add(buf, sizeof(struct usb_setup_packet));
```

Comment:
maybe
```c
net_buf_add_mem(buf, setup_packet, sizeof(struct usb_setup_packet));
```

Reply:
I don't really understand that case. You suggest that if `net_buf_add_mem` is used, we could avoid using the memcpy and save some CPU cycles, is this right? Or just a way to make code prettier?

Reply:
Just prettier, it combines to line into one